latex: hint that code-blocks continue on next page (closes #3764)

jfbu opened the pull request:
This adds new key for  ``'sphinxsetup'``. Provisory name is ``verbatimhintscontinuation``. 

The default is true and the effect is to print continuation hints when a code-block gets split in PDF output on multiple pages.

- Feature

However, if default value of  ``verbatimhintscontinuation`` was  ``false`` there would be backward compatibility, so it could be merged in stable.

### Relates
- #3764 

Currently the strings added to ``latex.tex_t`` have no translation. They are apart from an uppercase letter same as those from ``longtable.tex_t``. See #3791 in this context.

jfbu committed:
latex: hint that code-blocks continue on next page (closes #3764)

jfbu commented on doc/latex.rst:
```diff
@@ -218,6 +218,16 @@ Here are the currently available options together with their default values.
        before or after, but this is accessible currently only by re-defining some
        macros with complicated LaTeX syntax from :file:`sphinx.sty`.
 
+``verbatimhintscontinuation``
```
the name is not too good

jfbu commented on doc/latex.rst:
```diff
@@ -218,6 +218,16 @@ Here are the currently available options together with their default values.
        macros with complicated LaTeX syntax from :file:`sphinx.sty`.
 
+``verbatimhintscontinuation``
+    default ``true``. Boolean to specify if code blocks which happen to be split
```
with default to ``false`` we could possibly merge in stable, as it is then only refactoring of non api code, with same output

jfbu commented on sphinx/templates/latex/latex.tex_t:
```diff
@@ -30,6 +30,8 @@
 <%= numfig_format %>
 <%= pageautorefname %>
+\newcommand{\sphinxVerbatimContinuedText}{\footnotesize(<%= _('continued from previous page') %>)}
+\newcommand{\sphinxVerbatimContinuesText}{\footnotesize(<%= _('continues on next page') %>)}
```
as indicated in #3791 it seems gettext did not pick strings from longtable templates so I guess one will have to be careful about latex template as well

jfbu replied:
should I add string option to ``'sphinxsetup'`` such as ``verbatimcontinued``, ``verbatimcontinues`` ? with current code, user can use ``'preamble'`` key and ``\renewcommand``. With options, I need to then delay the definitions corresponding to the user options to at begin document. The latex file is then less understandable. Or, we move the lines above to before loading sphinx.sty

tk0miya replied:
I think .sty file is usually used for modify styles and macros in LaTeX world.
So I feel `preamble` key is better to do that.

Note: it might be better to allow user's .sty file instead of `preamble` key.
Now the key became much important, but it is boring to maintain it in conf.py.

jfbu commented on sphinx/texinputs/sphinx.sty:
```diff
@@ -256,6 +256,7 @@
 % verbatim
 \DeclareBoolOption[true]{verbatimwithframe}
 \DeclareBoolOption[true]{verbatimwrapslines}
+\DeclareBoolOption[true]{verbatimhintscontinuation}
```
see comment above about using rather ``false`` here

jfbu commented on sphinx/texinputs/sphinx.sty:
```diff
@@ -706,51 +707,80 @@
   \hskip\@totalleftmargin
   \hskip-\fboxsep\hskip-\fboxrule
-  #1{VerbatimBorderColor}{VerbatimColor}{#2}%
+  \spx@fcolorbox{VerbatimBorderColor}{VerbatimColor}{#1}{#2}{#3}%
```
formerly, special framing was used only to "glue" caption text to frame. Now we use it for continuation strings, hence always here `\spx@fcolorbox`. The extra parameters are the above and below text.

jfbu commented on sphinx/texinputs/sphinx.sty:
```diff
@@ -706,51 +707,80 @@
-\def\spx@fcolorbox #1#2%
-  {\color@b@x {\fboxsep\z@\color{#1}\spx@VerbatimFBox}{\color{#2}}}%
+\long\def\spx@fcolorbox #1#2#3#4%
+  {\color@b@x {\color{#1}\spx@VerbatimFBox{#3}{#4}}{\color{#2}}}%
```
formerly `\fboxsep` was set to zero, for the frame to touch the colored area

now, `\spx@VerbatimFBox` does not use `\fboxsep` so it is not needed to set it to zero

jfbu commented on sphinx/texinputs/sphinx.sty:
```diff
@@ -706,51 +707,80 @@
+\def\sphinxVerbatimLastFrameCommand
+   {\spx@colorbox\sphinxVerbatimContinued{}}
+
+\newcommand\sphinxVerbatimContinued{%
```
this macro maybe customized by knowledgeable LaTeX user

tk0miya commented:
I can't read the LaTeX macros well, so I will try to build PDF on my local later.
I will do it after 1.6.2 release. Please wait a moment :-)

jfbu commented:
sure I can wait ;-) a priori this is new feature hence for 1.7, although it could be merged into 1.6.x series if we set the default value of ``verbatimhintscontinuation`` to `false` (as a key for ``'sphinxsetup'``) so that it remains transparent and introduces no change by default.

Even if we introduce at 1.7 possibility to use  `tcolorbox` as discussed at #3790, I anticipate that it should be done as an option, keeping for a while the `framed.sty` use per default. Then this PR is not superfluous even if `tcolorbox` is usable at 1.7.

jfbu commented:
by the way, I forgot to mention the refactoring of LaTeX literal blocks in this PR makes possible now to get the code-block captions below rather than, as currently, above the frame.

tk0miya commented:
Did you remove borders of code-block? Personally I loved it.

jfbu commented:
To remove borders of code-block, you need

```
    'sphinxsetup': 'verbatimwithframe=false',
```

for details http://www.sphinx-doc.org/en/stable/latex.html#the-sphinx-latex-style-package-options

tk0miya commented:
In my local, the border is disabled by default. Even if I set `verbatimwithframe=true`.
Is this intended?

jfbu commented:
thanks for attentive review @tk0miya! this is not intended, and I confirmed it at my locale. I will fix that later... at some point I started testing with only background color. It worked fine in some earlier version with border. I must have done some latex bug afterwards. Thanks!

jfbu commented:
There was a missing brace pair at https://github.com/sphinx-doc/sphinx/pull/3792/commits/e1d7e3fdf7081a87ee5593cdd43218627cc88b0c#diff-f6432e57ae706b42894b186ffe699da0R727. As a result some  `\color` command from inside the box for the background color contaminated the border color. Thus the border was there but with color white :-(

jfbu commented on sphinx/texinputs/sphinx.sty:
```diff
@@ -706,56 +707,84 @@
-            % note that the caption brings \abovecaptionskip top vertical space
-            \moveright\dimexpr\fboxrule+.5\wd\@tempboxa
-                  \hb@xt@\z@{\hss\unhcopy\spx@VerbatimTitleBox\hss}\fi
+  \setbox\@tempboxa\hbox{{#3}}% contents with background color
```
The extra brace pair: `\hbox{{#3}}` is important because `#3` contains some `\color` command for the background. Without the extra braces, this colour contaminates the border. As the background is usually white, this means the border disappears.

tk0miya reviewed (approved):
Confirmed. LGTM!

jfbu commented:
Thanks for confirmation @tk0miya ;-).

1. I am not enthusiastic about my choice `verbatimhintscontinuation`. Perhaps `verbatimwithcontinuationhints` ? or `verbatimpleaseturnover` ([refs](http://www.macmillandictionary.com/thesaurus-category/british/abbreviations-in-letters-and-correspondence)), or `verbatimwithpto`...

2. If we set the default value of this latex boolean to `false`, then output is 100% identical as earlier. Exact definition of non public latex macros is modified, but I doubt Sphinx users had custom ones. And if they do, they know well TeX and can adapt... Thus perhaps then this could be merged in stable, (it does not change anything and makes customization possible) and our docs would document that default will change from `false` to `true` at 1.7 ? (or we leave it `false` always).

Do you have opinion on 1. and 2. ?

tk0miya commented:
1. Sorry, I don't have another idea
2. I agree to merge this into stable branch if it improves maintainability. You know we add new features to master branch basically, but it's okay to merge into stable if it prevents out development like merging operation (of cource, it should be stable enough)

jfbu commented:
Thanks for review @tk0miya I have rebased on stable

1. I changed `...hintscontinuation` into `..hintsturnover`. The former `continuation` was already used in macro names dealing with breaking of long code lines, so it was confusing.

2. I changed default of latex boolean to `false`, so as to be able to rebase on stable

3. I improved location of internationalized strings in latex document

4. inner refactoring to have maximal customizability.

In my testing with default `false` boolean, this does not change anything to documents; but it does modify non public macros in `sphinx.sty`.  Also, I will soon not be available for latex maintenance until September. Hence, in case of any doubt, it might be better to merge it in master for 1.7. Advantage of merging in 1.6.3 is that it will already be indicated in docs that `verbatimhintsturnover` boolean exists for `'sphinxsetup'`. At 1.7. it will be `true` by default, but now false. Disadvantage of merging in 1.6.3 is that some inner LaTeX macros have been changed not to fix bugs but to add potential feature.

jfbu commented:
For 1.7 I will make a PR which will allow to customize:
- position of code-block caption: before or after, left, center, or right
- position of continuation hints: left, center, or right.
In preparation for this I have amended this PR  to again give private names to some latex macros, because they will need to be changed at 1.7.

And I have added ``\sphinxstylecodecontinued``,  ``\sphinxstylecodecontinues`` to specify font or color of the continuation hints.

At 1.7, the horizontal positioning will be customized via a ``'sphinxsetup'`` option. The styling macros will need no change, they only decide of font and color and whether to use parentheses, etc, ... but not the position respective to code-block.

If no objection I will merge this to stable soon.

jfbu committed:
Move latex styling macros to proper place

jfbu commented:
Thanks for review @tk0miya I have now merged and will make a PR on master for caption position of literal blocks.
